File: src/ScOutfitter.Core/Routing.cs
using System.Text.Json.Nodes;

namespace ScOutfitter.Core;

/// <summary>Travel model of one quantum drive.</summary>
public sealed record QuantumDrive(string Name, double SpeedMps, double AccelMps2, double SpoolS, double CooldownS, double FuelPerMetre)
{
    public const double CalibrationS = 5.0;

    public static QuantumDrive From(Component c)
    {
        double speed = c.Stat("speed");
        double accel = c.Stat("accel");
        return new QuantumDrive(c.Name, speed, accel > 0 ? accel : speed / 10, c.Stat("spool"), c.Stat("cooldown"), c.Stat("fuel_rate"));
    }

    /// <summary>(seconds, fuel units) for one quantum jump of the given length.</summary>
    public (double Seconds, double Fuel) Hop(double distanceM)
    {
        if (distanceM <= 0)
        {
            return (0, 0);
        }

        double ramp = SpeedMps * SpeedMps / (2 * AccelMps2); // distance to reach full speed
        double travel = distanceM >= 2 * ramp
            ? 2 * SpeedMps / AccelMps2 + (distanceM - 2 * ramp) / SpeedMps
            : 2 * Math.Sqrt(distanceM / AccelMps2);
        return (SpoolS + CalibrationS + travel + CooldownS, distanceM * FuelPerMetre);
    }
}

public sealed record Buy(string Item, string Shop, int Quantity, int Price);

public sealed class Stop
{
    public required Location Location { get; init; }
    public List<Buy> Buys { get; init; } = [];
    public double LegSeconds { get; init; }
    public double LegFuel { get; init; }
    public double LegKm { get; init; }
    /// <summary>System jumps on the way here.</summary>
    public int Jumps { get; init; }
    /// <summary>false: not in the coordinate data, listed without distance or order.</summary>
    public bool Planned { get; init; } = true;
    public string System => Location.System;
    public int Cost => Buys.Sum(b => b.Price);
}

public sealed class Trip
{
    public required Location Start { get; init; }
    public List<Stop> Stops { get; init; } = [];
    /// <summary>Items not sold anywhere at all.</summary>
    public List<string> Unavailable { get; init; } = [];

    public IEnumerable<Stop> Planned => Stops.Where(s => s.Planned);
    public IEnumerable<Stop> Extra => Stops.Where(s => !s.Planned);
    public double Seconds => Planned.Sum(s => s.LegSeconds);
    public double Fuel => Planned.Sum(s => s.LegFuel);
    public double Km => Planned.Sum(s => s.LegKm);
    public int Cost => Stops.Sum(s => s.Cost);
}

/// <summary>Turns a shopping list into the shortest shopping trip through Stanton, Pyro and Nyx.</summary>
public static class Router
{
    /// <summary>Fixed overhead per stop: approach after quantum exit, landing, walk to the shop, take-off.</summary>
    private static double Overhead(LocationKind kind) => kind switch
    {
        LocationKind.Station => 180,
        LocationKind.City => 420,
        LocationKind.Outpost => 240,
        LocationKind.Gateway => 180,
        _ => 300,
    };

    /// <summary>Line up, enter and traverse a jump point.</summary>
    public const double JumpSeconds = 120;

    private sealed record Option(Location Location, string Shop, int UnitPrice);

    public static string FormatDuration(double seconds)
    {
        int total = (int)Math.Round(seconds);
        int h = total / 3600, m = total % 3600 / 60, s = total % 60;
        return h > 0 ? $"{h}h {m:00}m" : $"{m}m {s:00}s";
    }

    /// <summary>
    /// Picks the set of shops and the visiting order that minimizes travel time (plus price, if
    /// <paramref name="auecPerMinute"/> is set). The search space is small (a dozen shop locations,
    /// a handful of items), so subsets up to <paramref name="maxStops"/> and all orders are enumerated.
    /// Items only sold at places missing from the map are appended as unordered extra stops.
    /// </summary>
    public static Trip Plan(IReadOnlyList<Pick> picks, IReadOnlyDictionary<int, JsonNode> terminals, Starmap starmap,
        Location start, QuantumDrive qd, double auecPerMinute = 0, int maxStops = 5)
    {
        // what to buy, and where each item is on offer
        var needed = new Dictionary<string, int>(StringComparer.Ordinal);
        var comps = new Dictionary<string, Component>(StringComparer.Ordinal);
        foreach (Pick p in picks.Where(p => !p.Keep))
        {
            needed[p.Component.Name] = needed.GetValueOrDefault(p.Component.Name) + p.Quantity;
            comps[p.Component.Name] = p.Component;
        }

        var mapped = new Dictionary<string, Dictionary<string, Option>>(StringComparer.Ordinal);   // item -> location key -> option
        var unmapped = new Dictionary<string, Dictionary<(string, string), (string Shop, int Price)>>(StringComparer.Ordinal);
        var unavailable = new List<string>();
        foreach ((string name, Component comp) in comps)
        {
            var perLoc = new Dictionary<string, Option>(StringComparer.Ordinal);
            var perPlace = new Dictionary<(string, string), (string, int)>();
            foreach (Offer o in comp.Offers)
            {
                if (!terminals.TryGetValue(o.TerminalId, out JsonNode? term))
                {
                    continue;
                }

                string shop = term.Str("name") is { Length: > 0 } tn ? tn : o.TerminalName;
                Location? loc = starmap.LocateTerminal(term);
                if (loc is not null)
                {
                    if (!perLoc.TryGetValue(loc.Key, out Option? cur) || o.Price < cur.UnitPrice)
                    {
                        perLoc[loc.Key] = new Option(loc, shop, o.Price);
                    }
                }
                else
                {
                    (string, string) key = (term.Str("star_system_name") is { Length: > 0 } s ? s : "?", TerminalPlace(term));
                    if (!perPlace.TryGetValue(key, out (string Shop, int Price) cur) || o.Price < cur.Price)
                    {
                        perPlace[key] = (shop, o.Price);
                    }
                }
            }

            if (perLoc.Count > 0)
            {
                mapped[name] = perLoc;
            }
            else if (perPlace.Count > 0)
            {
                unmapped[name] = perPlace;
            }
            else
            {
                unavailable.Add(name);
            }
        }

        var stops = new List<Stop>();
        var locations = new Dictionary<string, Location>(StringComparer.Ordinal);
        foreach (Dictionary<string, Option> perLoc in mapped.Values)
        {
            foreach ((string key, Option opt) in perLoc)
            {
                locations[key] = opt.Location;
            }
        }

        (double Seconds, double Fuel, double Km, int Jumps) LegTo(Location a, Location b)
        {
            double secs = 0, fuel = 0, km = 0;
            int jumps = 0;
            foreach (Leg leg in starmap.Path(a, b))
            {
                if (leg.IsJump)
                {
                    secs += JumpSeconds;
                    jumps++;
                    continue;
                }

                double m = leg.From.DistanceKm(leg.To) * 1000;
                (double s, double f) = qd.Hop(m);
                secs += s;
                fuel += f;
                km += m / 1000;
            }

            return (secs + Overhead(b.Kind), fuel, km, jumps);
        }

        List<string> names = locations.Keys.OrderBy(k => k, StringComparer.Ordinal).ToList();
        double bestObjective = double.PositiveInfinity;
        List<Stop>? bestRoute = null;

        for (int n = 1; n <= Math.Min(maxStops, names.Count); n++)
        {
            foreach (string[] subset in Combinations(names, n))
            {
                if (!mapped.Keys.All(item => subset.Any(l => mapped[item].ContainsKey(l))))
                {
                    continue;
                }

                var buys = subset.ToDictionary(l => l, _ => new List<Buy>(), StringComparer.Ordinal);
                int priceTotal = 0;
                foreach ((string item, Dictionary<string, Option> options) in mapped)
                {
                    string l = subset.Where(options.ContainsKey).MinBy(x => options[x].UnitPrice)!;
                    Option opt = options[l];
                    int qty = needed[item];
                    buys[l].Add(new Buy(item, opt.Shop, qty, opt.UnitPrice * qty));
                    priceTotal += opt.UnitPrice * qty;
                }

                if (buys.Values.Any(b => b.Count == 0))
                {
                    continue; // a smaller subset covers the same shops
                }

                foreach (string[] order in Permutations(subset))
                {
                    var route = new List<Stop>();
                    Location prev = start;
                    double totalS = 0;
                    foreach (string l in order)
                    {
                        Location loc = locations[l];
                        (double secs, double fuel, double km, int jumps) = LegTo(prev, loc);
                        route.Add(new Stop { Location = loc, Buys = buys[l], LegSeconds = secs, LegFuel = fuel, LegKm = km, Jumps = jumps });
                        totalS += secs;
                        prev = loc;
                    }

                    // price is converted into minutes when the user values their time
                    double objective = totalS + (auecPerMinute > 0 ? priceTotal / auecPerMinute * 60 : 0);
                    if (objective < bestObjective)
                    {
                        bestObjective = objective;
                        bestRoute = route;
                    }
                }
            }
        }

        if (bestRoute is not null)
        {
            stops.AddRange(bestRoute);
        }

        // items only sold at places missing from the coordinate data: cheapest place per item
        var extra = new Dictionary<(string, string), Stop>();
        foreach ((string item, Dictionary<(string, string), (string Shop, int Price)> places) in unmapped)
        {
            ((string system, string place), (string shop, int price)) = places.MinBy(kv => kv.Value.Price);
            if (!extra.TryGetValue((system, place), out Stop? stop))
            {
                stop = new Stop { Location = new Location(place, system, place, 0, 0, 0, LocationKind.Station), Planned = false };
                extra[(system, place)] = stop;
            }

            int qty = needed[item];
            stop.Buys.Add(new Buy(item, shop, qty, price * qty));
        }

        stops.AddRange(extra.Values);
        return new Trip { Start = start, Stops = stops, Unavailable = unavailable };
    }

    private static string TerminalPlace(JsonNode term)
    {
        foreach (string f in new[] { "space_station_name", "city_name", "outpost_name", "moon_name", "planet_name", "orbit_name" })
        {
            string v = term.Str(f);
            if (v.Length > 0)
            {
                return v;
            }
        }

        return term.Str("name") is { Length: > 0 } n ? n : "?";
    }

    public static IEnumerable<string[]> Combinations(IReadOnlyList<string> items, int k)
    {
        int n = items.Count;
        if (k > n)
        {
            yield break;
        }

        int[] idx = Enumerable.Range(0, k).ToArray();
        while (true)
        {
            yield return idx.Select(i => items[i]).ToArray();
            int pos = k - 1;
            while (pos >= 0 && idx[pos] == n - k + pos)
            {
                pos--;
            }

            if (pos < 0)
            {
                yield break;
            }

            idx[pos]++;
            for (int j = pos + 1; j < k; j++)
            {
                idx[j] = idx[j - 1] + 1;
            }
        }
    }

    public static IEnumerable<string[]> Permutations(string[] items)
    {
        if (items.Length <= 1)
        {
            yield return items;
            yield break;
        }

        for (int i = 0; i < items.Length; i++)
        {
            string[] rest = items.Where((_, j) => j != i).ToArray();
            foreach (string[] tail in Permutations(rest))
            {
                yield return [items[i], .. tail];
            }
        }
    }
}

File: outfitter/routing.py
"""Turn a shopping list into the shortest shopping trip through Stanton."""
from __future__ import annotations

import itertools
import math
from dataclasses import dataclass, field

from .catalog import Component
from .optimizer import Pick
from .starmap import Location, Starmap

# Fixed overhead per stop: approach after quantum exit, landing, walking to the shop, take-off.
STOP_OVERHEAD_S = {"station": 180.0, "city": 420.0, "outpost": 240.0, "body": 300.0}
CALIBRATION_S = 5.0


@dataclass
class QuantumDrive:
    name: str
    speed: float      # m/s
    accel: float      # m/s^2 (stage two)
    spool: float      # s
    cooldown: float   # s
    fuel_rate: float  # fuel units per metre

    @classmethod
    def from_component(cls, c: Component) -> "QuantumDrive":
        s = c.stats
        return cls(c.name, s["speed"], s["accel"] or s["speed"] / 10, s["spool"], s["cooldown"], s["fuel_rate"])

    def hop(self, distance_m: float) -> tuple[float, float]:
        """(seconds, fuel units) for one quantum jump of distance_m."""
        if distance_m <= 0:
            return 0.0, 0.0
        d_ramp = self.speed ** 2 / (2 * self.accel)  # distance to reach full speed
        if distance_m >= 2 * d_ramp:
            travel = 2 * self.speed / self.accel + (distance_m - 2 * d_ramp) / self.speed
        else:
            travel = 2 * math.sqrt(distance_m / self.accel)
        return self.spool + CALIBRATION_S + travel + self.cooldown, distance_m * self.fuel_rate


@dataclass
class Buy:
    item: str
    shop: str
    quantity: int
    price: int  # total for the quantity


@dataclass
class Stop:
    location: Location
    buys: list[Buy] = field(default_factory=list)
    leg_seconds: float = 0.0
    leg_fuel: float = 0.0
    leg_km: float = 0.0
    system: str = "Stanton"
    planned: bool = True  # False: outside the coordinate data, listed without distance/order

    @property
    def cost(self) -> int:
        return sum(b.price for b in self.buys)


@dataclass
class Trip:
    start: Location
    stops: list[Stop]
    unavailable: list[str]  # items not sold anywhere at all

    @property
    def planned(self) -> list[Stop]:
        return [s for s in self.stops if s.planned]

    @property
    def extra(self) -> list[Stop]:
        return [s for s in self.stops if not s.planned]

    @property
    def seconds(self) -> float:
        return sum(s.leg_seconds for s in self.planned)

    @property
    def fuel(self) -> float:
        return sum(s.leg_fuel for s in self.planned)

    @property
    def km(self) -> float:
        return sum(s.leg_km for s in self.planned)

    @property
    def cost(self) -> int:
        return sum(s.cost for s in self.stops)


def _terminal_place(term: dict) -> str:
    for f in ("space_station_name", "city_name", "outpost_name", "moon_name", "planet_name", "orbit_name"):
        if term.get(f):
            return term[f]
    return term.get("name") or "?"


def _shop_options(picks: list[Pick], terminals: dict[int, dict], starmap: Starmap):
    """Per needed item: shops we can place on the map and shops we can only name.

    Returns (needed, mapped, unmapped, unavailable):
      needed   {item: quantity}
      mapped   {item: {location_name: (Location, shop, unit price)}}
      unmapped {item: {(system, place): (shop, unit price)}}
    """
    needed: dict[str, int] = {}
    comps: dict[str, Component] = {}
    for p in picks:
        if p.keep:
            continue
        needed[p.component.name] = needed.get(p.component.name, 0) + p.quantity
        comps[p.component.name] = p.component
    mapped: dict[str, dict[str, tuple[Location, str, int]]] = {}
    unmapped: dict[str, dict[tuple[str, str], tuple[str, int]]] = {}
    unavailable = []
    for name, comp in comps.items():
        per_loc: dict[str, tuple[Location, str, int]] = {}
        per_place: dict[tuple[str, str], tuple[str, int]] = {}
        for o in comp.offers:
            term = terminals.get(o.terminal_id)
            if not term:
                continue
            shop = term.get("name") or o.terminal_name
            loc = starmap.locate_terminal(term) if term.get("star_system_name") == "Stanton" else None
            if loc:
                if loc.name not in per_loc or o.price < per_loc[loc.name][2]:
                    per_loc[loc.name] = (loc, shop, o.price)
            else:
                key = (term.get("star_system_name") or "?", _terminal_place(term))
                if key not in per_place or o.price < per_place[key][1]:
                    per_place[key] = (shop, o.price)
        if per_loc:
            mapped[name] = per_loc
        elif per_place:
            unmapped[name] = per_place
        else:
            unavailable.append(name)
    return needed, mapped, unmapped, unavailable


def plan_trip(picks: list[Pick], terminals: dict[int, dict], starmap: Starmap, start: Location,
              qd: QuantumDrive, auec_per_minute: float = 0.0, max_stops: int = 5) -> Trip:
    """Pick the set of shops + visiting order that minimizes travel time (+ optional price weight).

    Small search space (a dozen shop locations, a handful of items), so we enumerate location
    subsets up to max_stops and all visiting orders. Items only sold outside the coordinate data
    (Pyro) are appended as unordered extra stops.
    """
    needed, mapped, unmapped, unavailable = _shop_options(picks, terminals, starmap)
    stops: list[Stop] = []

    locations: dict[str, Location] = {}
    for per_loc in mapped.values():
        for loc_name, (loc, _, _) in per_loc.items():
            locations[loc_name] = loc

    def leg(a: Location, b: Location) -> tuple[float, float, float]:
        d_m = a.distance_km(b) * 1000
        secs, fuel = qd.hop(d_m)
        return secs + STOP_OVERHEAD_S.get(b.kind, 300.0), fuel, d_m / 1000

    best: tuple[float, list[Stop]] | None = None
    names = sorted(locations)
    for n in range(1, min(max_stops, len(names)) + 1):
        for subset in itertools.combinations(names, n):
            if not all(any(l in mapped[item] for l in subset) for item in mapped):
                continue
            buys: dict[str, list[Buy]] = {l: [] for l in subset}
            price_total = 0
            for item in mapped:
                l = min((l for l in subset if l in mapped[item]), key=lambda l: mapped[item][l][2])
                _, shop, price = mapped[item][l]
                qty = needed[item]
                buys[l].append(Buy(item, shop, qty, price * qty))
                price_total += price * qty
            if any(not b for b in buys.values()):
                continue  # a smaller subset covers the same shops
            for order in itertools.permutations(subset):
                route: list[Stop] = []
                prev = start
                total_s = 0.0
                for l in order:
                    loc = locations[l]
                    secs, fuel, km = leg(prev, loc)
                    route.append(Stop(loc, buys[l], secs, fuel, km))
                    total_s += secs
                    prev = loc
                # price is converted into minutes when the user values their time
                objective = total_s + (price_total / auec_per_minute * 60 if auec_per_minute else 0)
                if best is None or objective < best[0]:
                    best = (objective, route)
    if best:
        stops.extend(best[1])

    # items only sold where we have no coordinates: cheapest place per item, grouped by place
    extra: dict[tuple[str, str], Stop] = {}
    for item, places in unmapped.items():
        (system, place), (shop, price) = min(places.items(), key=lambda kv: kv[1][1])
        stop = extra.setdefault((system, place), Stop(Location(place, place, (0, 0, 0), "station"),
                                                      system=system, planned=False))
        qty = needed[item]
        stop.buys.append(Buy(item, shop, qty, price * qty))
    stops.extend(extra.values())
    return Trip(start, stops, unavailable)


def fmt_duration(seconds: float) -> str:
    m, s = divmod(int(round(seconds)), 60)
    h, m = divmod(m, 60)
    return f"{h}h {m:02d}m" if h else f"{m}m {s:02d}s"
